src/udp/connection/encoded_id.rs

```diff
@@ -0,0 +1,346 @@
+        after_now: Option<Duration>,
+    ) -> Result<(), Self::Error> {
+        if Self::new_int(client_id.clone(), lifetime, false, after_now).value == encoded_id {
+            println!("Now");
+            return Ok(());
+        } else if Self::new_int(client_id.clone(), lifetime, true, after_now).value == encoded_id {
+            println!("Past");
+            return Ok(());
+        } else {
+            return Err("Expired connection id");
+        }
+    }
+}
+
+impl EncodedId<KeyedClientId> for WitnessEncodedId {
+    type Error = &'static str;
+
+    fn new(client_id: KeyedClientId, lifetime: Duration) -> Self {
+        let expiry_u32: u32 = LocalClock::next_as_timestamp(&lifetime).try_into().unwrap();
+
+        let mut hasher = DefaultHasher::default();
+        hasher.write_u32(expiry_u32);
+        hasher.write(client_id.value().as_slice());
+        let witness = hasher.finish().to_le_bytes();
+
+        let id: Vec<u8> = [&witness[0..4], expiry_u32.to_le_bytes().as_slice()].concat();
+
+        WitnessEncodedId {
+            value: id.try_into().unwrap(),
+        }
+    }
+
+    fn value(&self) -> &[u8; 8] {
+        &self.value
+    }
+
+    fn check(
+        client_id: KeyedClientId,
+        _lifetime: Duration,
+        encoded_id: [u8; 8],
+        after_now: Option<Duration>,
+    ) -> Result<(), Self::Error> {
+        let expiry_bytes: [u8; 4] = encoded_id[4..8].try_into().unwrap();
+        let expiry_u32 = u32::from_le_bytes(expiry_bytes);
+
+        let after_maybe = match after_now {
+            Some(after) => after,
+            None => Duration::ZERO,
+        };
+
+        let now_u32 = LocalClock::next_as_timestamp(&after_maybe).try_into().unwrap();
+
+        if expiry_u32 <= now_u32 {
+            return Err("Expired connection id");
+        }
+
+        let mut hasher = DefaultHasher::default();
+        hasher.write_u32(expiry_u32);
+        hasher.write(client_id.value().as_slice());
+
+        let witness = hasher.finish().to_le_bytes();
+
+        if &witness[0..4] != &encoded_id[0..4] {
+            Err("Bad Witness in Connection Id")
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl EncodedId<PlainClientId> for EncryptedEncodedId {
+    type Error = &'static str;
+
+    fn new(client_id: PlainClientId, lifetime: Duration) -> Self {
+        let expiry_u32: u32 = LocalClock::next_as_timestamp(&lifetime).try_into().unwrap();
+
+        let id_clear: Vec<u8> = [&client_id.value()[0..4], &expiry_u32.to_le_bytes().as_slice()].concat();
+
+        let id_clear_bytes: [u8; 8] = id_clear.try_into().unwrap();
+        let mut block: BlowfishArray = BlowfishArray::from(id_clear_bytes);
+
+        let blowfish_key: &blowfish::Blowfish<LE> = if cfg!(test) { &TEST_BLOWFISH_KEY } else { &BLOWFISH_KEY };
+
+        <Blowfish<LE> as BlockEncrypt>::encrypt_block(blowfish_key, &mut block);
+
+        EncryptedEncodedId {
+            value: block.try_into().unwrap(),
+        }
+    }
```

Comment:
```rust
fn new(client_id: PlainClientId, lifetime: Duration) -> Self {

    // 1.- Generate connection id data
    let expiry_u32: u32 = LocalClock::next_as_timestamp(&lifetime).try_into().unwrap();
    let id_clear: Vec<u8> = [&client_id.value()[0..4], &expiry_u32.to_le_bytes().as_slice()].concat();
    let id_clear_bytes: [u8; 8] = id_clear.try_into().unwrap();

    // 2.- Encrypt connection id data
    let mut block: BlowfishArray = BlowfishArray::from(id_clear_bytes);
    let blowfish_key: &blowfish::Blowfish<LE> = if cfg!(test) { &TEST_BLOWFISH_KEY } else { &BLOWFISH_KEY };
    <Blowfish<LE> as BlockEncrypt>::encrypt_block(blowfish_key, &mut block);

    EncryptedEncodedId {
        value: block.try_into().unwrap(),
    }
}
```

@da2ce7 

- I suppose we do not need the suffix here: `expiry_u32`
- I do not like the Rust verbosity just to concat two bytes arrays. I've been trying to get rid of it by extracting methods.

```rust
pub IdClear {
    bytes: [u8; 4];
    
    fn new(client_id: &PlainClientId, expiry: u32) -> Self {
      let client_id_witness = &client_id.value()[0..4];
      let expiry_as_bytes = expiry_u32.to_le_bytes();
      let id_clear: Vec<u8> = [client_id_witness, expiry_as_bytes].concat();
      let bytes: [u8; 8] = id_clear.try_into().unwrap();
      Self {
          bytes 
      }
    }

    fn bytes(&self) -> [u8; 8] {
        self.bytes
    }
}

fn new(client_id: PlainClientId, lifetime: Duration) -> Self {

    // 1.- Generate connection id data
    let expiry: u32 = LocalClock::next_as_timestamp(&lifetime).try_into().unwrap();
    let id_clear_bytes: [u8; 8] = IdClear::new(client_id, expiry).bytes()

    // 2.- Encrypt connection id data
    let mut block: BlowfishArray = BlowfishArray::from(id_clear_bytes);
    let blowfish_key: &blowfish::Blowfish<LE> = if cfg!(test) { &TEST_BLOWFISH_KEY } else { &BLOWFISH_KEY };
    <Blowfish<LE> as BlockEncrypt>::encrypt_block(blowfish_key, &mut block);

    EncryptedEncodedId {
        value: block.try_into().unwrap(),
    }
}
```

- As I mentioned before, I think this `if cfg!(test) { &TEST_BLOWFISH_KEY } else { &BLOWFISH_KEY };` should be moved out of production code.